tests-ng/plugins/sysdiff.py

```diff
@@ -0,0 +1,482 @@
+
+                filtered_files = sorted([
+                    f for f in files
+                    if not self.should_ignore_file(f, ignore_patterns)
+                ])
+
+                for filepath in filtered_files:
+                    result = self.shell(f'sha256sum -- "{filepath}" 2>/dev/null', capture_output=True, ignore_exit_code=True)
+                    if result.stdout.strip():
+                        hash_part = result.stdout.strip().split()[0]
+                        file_hashes[filepath] = hash_part
+
+            finally:
+                os.unlink(paths_file)
+
+        except Exception as e:
+            print(f"Error collecting file hashes: {e}")
+
+        return file_hashes
+
+
+class SnapshotManager:
+    """Manages snapshot creation, storage, and retrieval"""
+
+    def __init__(self, state_dir: Path = None):
+        self.state_dir = state_dir or STATE_DIR
+        self.state_dir.mkdir(parents=True, exist_ok=True)
+
+    def create_snapshot(self, name: str = None, paths: List[str] = None,
+                       ignore_file: Path = None) -> Snapshot:
+        """Create a new system snapshot"""
+        if paths is None:
+            paths = DEFAULT_PATHS
+
+        if name is None:
+            name = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+
+        snapdir = self.state_dir / name
+        if snapdir.exists():
+            raise ValueError(f"Snapshot '{name}' already exists at {snapdir}")
+
+        snapdir.mkdir(parents=True)
+
+        shell = ShellRunner(None)
+        dpkg = Dpkg(shell)
+        systemd = Systemd(shell)
+        file_collector = FileCollector(shell)
+        sysctl_collector = Sysctl(shell)
+        kernel_module = KernelModule(shell)
+
+        packages = dpkg.collect_packages()
+        systemd_units = systemd.list_units()
+        sysctl_params = sysctl_collector.collect_sysctl_parameters()
+        kernel_modules = kernel_module.collect_loaded_modules()
+        ignore_patterns = file_collector.load_ignore_patterns(ignore_file)
+        normalized_paths = file_collector.normalize_paths(paths)
+        files = file_collector.collect_file_hashes(normalized_paths, ignore_patterns)
+
+        metadata = SnapshotMetadata(
+            created_at=datetime.now().isoformat(),
+            paths=normalized_paths,
+            ignore_file=bool(ignore_file and ignore_file.exists())
+        )
+
+        snapshot = Snapshot(name, metadata, packages, systemd_units, files, sysctl_params, kernel_modules)
+        self._save_snapshot(snapshot, snapdir)
+
+        return snapshot
+
+    def _save_snapshot(self, snapshot: Snapshot, snapdir: Path):
+        """Save snapshot to disk"""
+
+        with open(snapdir / "packages.txt", 'w') as f:
+            for package, version in snapshot.packages.items():
+                f.write(f"{package}\t{version}\n")
+
+        with open(snapdir / "systemd_units.txt", 'w') as f:
```

Comment:
would it be better to use json here? I see there is an unused json import, so maybe you wanted to do this and found an issue with that? this tab-separated format does not seem very stable to me.